pkg/proxy/apis/config/v1alpha2/defaults.go

```diff
@@ -0,0 +1,135 @@
+		obj.OOMScoreAdj = &temp
+	}
+	if obj.IPTables.SyncPeriod.Duration == 0 {
+		obj.IPTables.SyncPeriod = metav1.Duration{Duration: 30 * time.Second}
+	}
+	if obj.IPVS.SyncPeriod.Duration == 0 {
+		obj.IPVS.SyncPeriod = metav1.Duration{Duration: 30 * time.Second}
+	}
+	zero := metav1.Duration{}
+	if obj.UDPIdleTimeout == zero {
+		obj.UDPIdleTimeout = metav1.Duration{Duration: 250 * time.Millisecond}
+	}
+
+	if obj.Conntrack.MaxPerCore == nil {
+		obj.Conntrack.MaxPerCore = pointer.Int32Ptr(32 * 1024)
+	}
+	if obj.Conntrack.Min == nil {
+		obj.Conntrack.Min = pointer.Int32Ptr(128 * 1024)
+	}
+
+	if obj.IPTables.MasqueradeBit == nil {
+		temp := int32(14)
+		obj.IPTables.MasqueradeBit = &temp
+	}
+	if obj.Conntrack.TCPEstablishedTimeout == nil {
+		obj.Conntrack.TCPEstablishedTimeout = &metav1.Duration{Duration: 24 * time.Hour} // 1 day (1/5 default)
+	}
+	if obj.Conntrack.TCPCloseWaitTimeout == nil {
+		// See https://github.com/kubernetes/kubernetes/issues/32551.
+		//
+		// CLOSE_WAIT conntrack state occurs when the Linux kernel
+		// sees a FIN from the remote server. Note: this is a half-close
+		// condition that persists as long as the local side keeps the
+		// socket open. The condition is rare as it is typical in most
+		// protocols for both sides to issue a close; this typically
+		// occurs when the local socket is lazily garbage collected.
+		//
+		// If the CLOSE_WAIT conntrack entry expires, then FINs from the
+		// local socket will not be properly SNAT'd and will not reach the
+		// remote server (if the connection was subject to SNAT). If the
+		// remote timeouts for FIN_WAIT* states exceed the CLOSE_WAIT
+		// timeout, then there will be an inconsistency in the state of
+		// the connection and a new connection reusing the SNAT (src,
+		// port) pair may be rejected by the remote side with RST. This
+		// can cause new calls to connect(2) to return with ECONNREFUSED.
+		//
+		// We set CLOSE_WAIT to one hour by default to better match
+		// typical server timeouts.
+		obj.Conntrack.TCPCloseWaitTimeout = &metav1.Duration{Duration: 1 * time.Hour}
+	}
+	if obj.ConfigSyncPeriod.Duration == 0 {
+		obj.ConfigSyncPeriod.Duration = 15 * time.Minute
+	}
+
+	if len(obj.ClientConnection.ContentType) == 0 {
+		obj.ClientConnection.ContentType = "application/vnd.kubernetes.protobuf"
+	}
+	if obj.ClientConnection.QPS == 0.0 {
+		obj.ClientConnection.QPS = 5.0
+	}
+	if obj.ClientConnection.Burst == 0 {
+		obj.ClientConnection.Burst = 10
+	}
+	if obj.FeatureGates == nil {
+		obj.FeatureGates = make(map[string]bool)
+	}
+}
+
+// getDefaultAddresses returns default address of healthz and metrics server
+// based on the given bind address. IPv6 addresses are enclosed in square
+// brackets for appending port.
+func getDefaultAddresses(bindAddress string) (defaultHealthzAddress, defaultMetricsAddress string) {
+	if net.ParseIP(bindAddress).To4() != nil {
+		return "0.0.0.0", "127.0.0.1"
+	}
+	return "[::]", "[::1]"
```

Comment:
question, do we need to use square brackets for the health and metrics addressed using IPv6 ?
if affirmative, should be better return the address without square brackets and use https://golang.org/pkg/net/#JoinHostPort to create the URL?

Reply:
I think so - I copied that from the v1alpha1, but there may not have been a strong reason for it in the first place. Though looking at that, that I also need to revisit the logic... it should be checking what the bindAddresses are.